### ISSUE TITLE
nD Visulization

### ISSUE THREAD

**Issue by max0410** [Smilebags/p5.dimensions.js#24]
Looking into the code of this [website](http://www.math.union.edu/~dpvc/math/4d/models/) I see it uses a 3d javascript library called X3D. The things they are showing are 3d projections of 4d objects. For example, are eyes can not see 3d, they see 2d. So, our life is a 2d projection of a 3d world. Pretty deep right. So for in order for the nD Visualization to work, we will have to take our nD data and project it on to a 3d canvas, most likely using p5js's tool for 3d objects. So we just have to figure out how to project nD objects to 3d. I have been researching on this.

2d projection of 0d hypercube
![image](https://cloud.githubusercontent.com/assets/16921177/23090764/e8cfe0d8-f573-11e6-864d-0560d11520cc.png)

2d projection of 1d hypercube
![image](https://cloud.githubusercontent.com/assets/16921177/23090798/a1a2b702-f574-11e6-8978-e379feac84d4.png)

2d projection of 2d hypercube
![image](https://cloud.githubusercontent.com/assets/16921177/23090776/22d7246c-f574-11e6-97a9-86e465ff98c8.png)

2d projection of 3d hypercube
![image](https://cloud.githubusercontent.com/assets/16921177/23090782/58843f78-f574-11e6-86f7-30596b4e1a7c.png)

2d projection of 4d hypercube
![image](https://cloud.githubusercontent.com/assets/16921177/23090785/65538b50-f574-11e6-8118-e90dc46d92e1.png)

2d projection of 5d hypercube
![image](https://cloud.githubusercontent.com/assets/16921177/23090791/718a170e-f574-11e6-944d-5af72ee42268.png)

3d projection of a 4d hypercube
http://codepen.io/max0410/pen/zNQQQM

**Comment by Smilebags** [comment]
Great work. Yes, projection matrices are what we will need for that. A projection matrix for 3>2 might look like this:
`(x',y') = (x/z, y/z)`
So, what we need to do is the same thing from 4 to 3. I think the best way to do this will to be first implement a matrix function which can take a nVector and another matrix argument, and return the resulting nVector. Are you familiar with matrices? If so would you like to start implementing a function for doing that?

Maybe we should define an nMatrix first, as a standard which we can always base the functions off. I'll work on that now.

**Comment by max0410** [comment]
![image](https://cloud.githubusercontent.com/assets/16921177/23090295/201499ea-f569-11e6-8ace-e3b46b5610e6.png)
Here is a crappy drawing of 4d cube

**Comment by Smilebags** [comment]
Yeah, that looks right. Well, it is a 2D projection of a 3D projection of a 4D cube.

**Comment by max0410** [comment]
I was just about to post a 2d projection of a 5d object
If we get the 2d projections down, then we ccan do the 3d ones

**Comment by Smilebags** [comment]
Yep. I just committed an initial idea for how to implement matrices, but not anywhere near complete.

**Comment by max0410** [comment]
![image](https://cloud.githubusercontent.com/assets/16921177/23090350/7bc32dc8-f56a-11e6-9cdc-b6d8ae67c328.png)

5d projection
My brain hurts now

**Comment by max0410** [comment]
Ok, I'm going to figure out p5js's 3d stuff.

**Comment by Smilebags** [comment]
That would be great. There's quite a lot there.

**Comment by max0410** [comment]
Here is a 3d projection of a 4d wireframe. http://codepen.io/max0410/pen/xgNWjG

**Comment by max0410** [comment]
I used custom shapes so I could create just a wireframe, I am trying to find a way to just use box(), and then have translucent faces.

**Comment by Smilebags** [comment]
Translucency is pretty complex in simple engines like webGL, you can look it up but it might be difficult.

**Comment by Smilebags** [comment]
Nice work with your code, it looks good!

**Comment by max0410** [comment]
Thanks!

**Comment by max0410** [comment]
I created an adjustable 4d hypercube: http://codepen.io/max0410/pen/zNQQQM
Now I will use the `hypercube4d()` function to create a 5d hypercube!

**Comment by Smilebags** [comment]
Looks good! Check out my new example and see if you can extend it. :)
